### PR TITLE
Change hover cards to not appear until the mouse stops in web UI

### PR DESCRIPTION
**Pull request by Gargron**
- Reduce the amount of times that events get added and removed from body because of state changes
- Block hover cards from appearing when actively scrolling
- Delay hover cards until mouse movement stops (entirely)

### CODE REVIEW

**Comment by shleeable** [comment]
Still glitching on my firefox..
[screen-recorder-thu-jun-27-2024-14-16-03.webm](https://github.com/mastodon/mastodon/assets/17537000/a317f513-4415-4494-80ff-85de7526b661)

**Comment by shleeable** [comment]
Change to another tab.. click around.. tab back to mastodon and the hover is just there for the user is just there.
image

**Comment by vmstan** [comment]
I think all of the feedback I've been getting from users so far that has been negative has been from those who are in the advanced interface. I'm wondering if it's even necessary there when a click already brings up their profile in a column.

File: app/javascript/mastodon/components/hover_card_controller.tsx
Comment (renchap) [inline app/javascript/mastodon/components/hover_card_controller.tsx:94]: I am not sure if I understand it correctly, but you never cancel the previous scroll timeout, so it will trigger?

I have a feeling you want a debounce here, to call `scrollingRef.current = false` 100ms after the last scroll? If yes, maybe you could use the https://github.com/xnimorz/use-debounce package that I introduced in the grouped notifications PR?

But it would not be compatible with my idea above of not having it in a ref.

File: app/javascript/mastodon/components/hover_card_controller.tsx
Comment (renchap) [inline app/javascript/mastodon/components/hover_card_controller.tsx:28]: Maybe rename to `isScrollingRef`?

Also does it need to be defined as a ref? I wonder if this could be a `const` in `useEffect`, as it is only used there.

**Comment by rsholmes** [comment]
When scrolling the feed (I use the advanced interface) my mouse isn't moving at all, rather the feed is, and if I happen to stop scrolling when a user link is under or near the mouse, up pops the unwanted popup. This happens a lot, and it is exasperating.

I can confirm behavior reported by shleeable, just had one of these pop up with the cursor (present and recent past location) about 20% of the width of the screen away. VERY annoying.

I also have seen them fail to go away when the cursor moves well outside their range (though they usually do). Here for instance the cursor was on the elephant's paper airplane when I took the screen shot:

image

vmstan writes:

> I'm wondering if it's even necessary [in the advanced interface] when a click already brings up their profile in a column.

Very good point. I would wholeheartedly support keeping this popup out of the advanced interface.

File: app/javascript/mastodon/components/hover_card_controller.tsx
Comment (Gargron) [reply app/javascript/mastodon/components/hover_card_controller.tsx:124]: `setScrollTimeout` cancels the previous timeout if there is one. Basically, 100ms after the last scroll event `scrollingRef.current` will be set to `false`.

**Comment by maikelthedev** [comment]
This needs to be "disableable" on the advance Web interface at least. It completely destroyed my desktop user experience to the point I'm now only using someone else's android client until this annoyance gets out of the way.

**Comment by vmstan** [comment]
> This needs to be disableable on the advance Web interface at least. It completely destroyed my desktop user experience to the point I'm now only using someone else's android client until this annoyance gets out of the way.

@maikelthedev I've just pushed the latest version of this PR to vmst.io and it's a big improvement from what I've seen so far.

**Comment by ocdtrekkie** [comment]
@maikelthedev It's really worth highlighting that if you are using experimental code, your experience might not reflect the final one! Ideally if it's annoying now, the goal should not be to shut it off, but to fix it so it is not annoying before it is released as a feature.

**Comment by maikelthedev** [comment]
>  the goal should not be to shut it off, 

Disagree, it should be disableable. As others have mentioned elsewhere, it's entirely unneeded on the advance web interface. You can just click on any profile to have all that info shown, without it taking space of your timelines. If we want to shut if off we should be able to do so. 

@vmstan 
> I've just pushed the latest version of this PR to vmst.io and it's a big improvement from what I've seen so far.

Not entirely yet there, it does improve *yes* but it is still very glitchy, it has now a delay, sure, and it allows to scroll, UNTIL you rest the mouse and that delay is GONE, and suddenly it shows all over the place whenever your mouse goes. go around the screen, the hover cards follow you all over the place, it reminds me to those old cursor effects that left a trail behind, like this. 

image

Only that is not trail of the mouse, it is the hover cards chasing you.

**Comment by rsholmes** [comment]
The UI assumption here is that when the mouse is stationary over a user link, it's because the user wanted to put it on that link. This is a faulty assumption. It may be there because the link was scrolled there; it may be there because the user needed to move the mouse away from somewhere else, e.g. to unobstruct some text. With something like tooltips you can often get away with these assumptions because tooltips are small and don't obstruct much, and the icons the tooltips are associated with often are also small and stationary and placed near the edges of the windows. These popups obstruct a lot, and the user links that trigger them move around when scrolling and occupy a significant fraction of the window.

The result is behavior many users are finding detrimental to the user experience, and there are no simple fixes, e.g. waiting for the mouse to stop doesn't help if you are scrolling or moving the mouse away from something else.

Given all that, the developers do indeed need to choose either to abandon the whole effort, or mitigate it by not having the popups in the advanced interface where they are unneeded, or add a setting to disable the popups for users who do not want to be subjected to them.

**Comment by maikelthedev** [comment]
I couldn't agree more with @rsholmes . Let me add some extra info:

I've seen this before in Misskey-based activity pub implementations. But there this is not obtrusive in neither the simple interface nor the Deck mode (there the advance web interface is deck mode). This is how I've seen it operate there:

1. It has the delay
2. The delay resets once the pop-up closes (this is the bit that isn't currently working, the delay is not resetting). 
3. Very few elements trigger the pop-up. 

So I can scroll and scroll, unbothered ever. 

IMHO I wouldn't scrap it, it is a good idea. I think what it needs is:
1. The delay to reset when it hides otherwise we're back to square one with it being an obstacle chasing the cursor. 
2. LESS elements triggering it. Please do test moving around especially in the advance web interface where you do not need the clutter it provides. 
3. A settings in INSTANCENAME/settings/preferences/appearance to be able to entirely disable it. I can imagine it might also be a problem with people with screen readers.

**Comment by MadokVaur** [comment]
> The result is behavior many users are finding detrimental to the user experience, and there are no simple fixes, e.g. waiting for the mouse to stop doesn't help if you are scrolling or moving the mouse away from something else.
> 
> Given all that, the developers do indeed need to choose either to abandon the whole effort, or mitigate it by not having the popups in the advanced interface where they are unneeded, or add a setting to disable the popups for users who do not want to be subjected to them.

I've been following this closely, and as a pure Mastodon user (not a coder) exclusively of the web UI, what I'm seeing is a UX/UI "enhancement" that makes the web version a minefield for the mouse cursor and the unwary user

Acute attention will need to be paid to where the cursor comes to rest at any given moment

And two core questions never seem to have been asked (that I've seen) let alone answered:

Who decided that popup user profiles would be a good idea in the first place, and how many times do people need to see the same unchanged user profile pop up, again and again and again when user profiles almost never change, only because the user was careless about where the mouse cursor came to rest at any given moment?

**Comment by Gargron** [comment]
> Who decided that popup user profiles would be a good idea in the first place

#14918

**Comment by MadokVaur** [comment]
> > Who decided that popup user profiles would be a good idea in the first place
> 
> #14918

Fair

And the second half of my question:

"...and how many times do people need to see the same unchanged user profile pop up, again and again and again when user profiles almost never change, only because the user was careless about where the mouse cursor came to rest at any given moment?"

**Comment by shleeable** [comment]
> This feature is so horrible it killed my parents`.

Please add a toggle to disable the popup in the settings...

twitter/elk and a bunch of other clients have hover popups and I've never accidentally triggers the user hover on twitter... so it's just a matter of polish at this point.

**Comment by cadusilva** [comment]
It's getting better. Just tried the latest patch and it's more polished than the first batch. I'm in favour of letting the user decide if they want to see the pop-ups via settings option, but also about leaving it in the advanced interface too.

**Comment by renchap** [comment]
Everyone, please keep the comments here constructive. This is a brand new feature and we know it needs some tuning.

Your feedback is very welcome and we are taking it into account to adjust some things and make it behave correctly.

**Comment by shleeable** [comment]
I just remembered I'm running a custom theme `bird-ui`.. and that might be the cause of my glitches.

@vmstan or anybody running this code.. can you test on stock vs theme?

**Comment by vmstan** [comment]
All my testing has been with the stock theme.

**Comment by maikelthedev** [comment]
> All my testing has been with the stock theme.

same

**Comment by shleeable** [comment]
I've reset my firefox profile and used the recent commit... and my problems have stopped

I'll continue to test and see if anything pops up, but my annoyances seem to have been cyber gremlins

**Comment by shleeable** [comment]
Everything is looking great. Merge away!

[screen-recorder-mon-jul-01-2024-14-50-53.webm](https://github.com/mastodon/mastodon/assets/17537000/a260515a-2be9-42a3-898e-a3b31baf7f28)

**Comment by maikelthedev** [comment]
I've done
1. Force refreshed the page
2. Move the mouse around the minefield. 
3. Activated a hover card on purpose 
4. Moved the mouse around the other HC activators. 

The time resets. Not a trace of hover cards following my mouse. Yay! 🥳 

I think you guys polished it. It works the same as Misskey now. 

I still think there's one very needed improvement though: the ability to go to settings/appearance and entirely disabled it.

File: app/javascript/mastodon/components/hover_card_controller.tsx
Comment (ClearlyClaire) [inline app/javascript/mastodon/components/hover_card_controller.tsx:1]: I would not be against a second pair of eyes on this file, @renchap

**Review by ClearlyClaire — APPROVED** [review]
I wouldn't be against having another pair of eyes on the more technical parts of the event handling lifecycle in `HoverCardController`, but it looks good to me overall.

A quick field test has surfaced none of the bug I had seen up until now, and everything seems to work very well. If anything, I think the `enterDelay` might be a bit long now.

I agree with other commenters that a setting to disable the feature entirely could be useful, but I don't think it's absolutely necessary either, and I don't think it belongs in this PR, it can be done later.

**Review by renchap — APPROVED** [review]
Looks good to me.

I would prefer a different `useTimeout(callback)` hook that does not try to memoize it, but takes a memoized callback as an argument, but this can be changed later.

**Comment by MattWilcox** [comment]
Would love an ability to disable this feature please - I didn't know about it until I accedentally triggered it a few times, and it's purely getting in my way, without me ever wanting what it does. Thanks.

https://mstdn.social/@mattwilcox/112734403770084644

**Comment by rsholmes** [comment]
I've installed the Stylus Firefox extension and use it to impose
```
    .hover-card {
        display:none
    }
```
A disable option would be preferable but this works for now.

**Comment by vmstan** [comment]
@rsholmes now hacks required now after #30931 was merged

image

**Comment by MattWilcox** [comment]
great stuff, thanks for this change :D

**Pull request merged**
